fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/dynamic/DynamicSchema.java

```diff
@@ -0,0 +1,434 @@
+        Set<String> allFdProtoNames = collectFileDescriptorNamesAndCheckForDupes(fileDescSet);
+
+        // build FileDescriptors, resolve dependencies (imports) if any
+        Map<String, FileDescriptor> resolvedFileDescMap = new HashMap<>();
+        while (resolvedFileDescMap.size() < fileDescSet.getFileCount()) {
+            for (FileDescriptorProto fdProto : fileDescSet.getFileList()) {
+                if (resolvedFileDescMap.containsKey(fdProto.getName())) {
+                    continue;
+                }
+
+                List<String> dependencyList = fdProto.getDependencyList();
+                List<FileDescriptor> resolvedFdList = new ArrayList<>();
+                for (String depName : dependencyList) {
+                    if (!allFdProtoNames.contains(depName)) {
+                        throw new IllegalArgumentException("cannot resolve import " + depName + " in " + fdProto.getName());
+                    }
+                    FileDescriptor fd = resolvedFileDescMap.get(depName);
+                    if (fd != null) {
+                        resolvedFdList.add(fd);
+                    }
+                }
+
+                if (resolvedFdList.size() == dependencyList.size()) { // dependencies resolved
+                    FileDescriptor[] fds = new FileDescriptor[resolvedFdList.size()];
+                    FileDescriptor fd = FileDescriptor.buildFrom(fdProto, resolvedFdList.toArray(fds));
+                    resolvedFileDescMap.put(fdProto.getName(), fd);
+                }
+            }
+        }
+
+        return resolvedFileDescMap;
+    }
+
+    private static Set<String> collectFileDescriptorNamesAndCheckForDupes(@Nonnull final FileDescriptorSet fileDescSet) {
+        Set<String> result = new HashSet<>();
+        for (FileDescriptorProto fdProto : fileDescSet.getFileList()) {
+            if (result.contains(fdProto.getName())) {
+                throw new IllegalArgumentException(String.format(duplicateNameErrorMessage, fdProto.getName()));
+            }
+            result.add(fdProto.getName());
+        }
+        return result;
+    }
+
+    private void addMessageType(Descriptor msgType, String scope, Set<String> msgDupes, Set<String> enumDupes) {
+        String msgTypeNameFull = msgType.getFullName();
+        String msgTypeNameShort = (scope == null ? msgType.getName() : scope + "." + msgType.getName());
+
+        if (msgDescriptorMapFull.containsKey(msgTypeNameFull)) {
+            throw new IllegalArgumentException(String.format(duplicateNameErrorMessage, msgTypeNameFull));
+        }
+        if (msgDescriptorMapShort.containsKey(msgTypeNameShort)) {
+            msgDupes.add(msgTypeNameShort);
+        }
+
+        msgDescriptorMapFull.put(msgTypeNameFull, msgType);
+        msgDescriptorMapShort.put(msgTypeNameShort, msgType);
+
+        for (Descriptor nestedType : msgType.getNestedTypes()) {
+            addMessageType(nestedType, msgTypeNameShort, msgDupes, enumDupes);
```

Comment:
Can there ever be an infinite loop where two types have a circular reference?

Reply:
Great point, I just checked and it looks like PB allows recursive message definitions.

I think this method will not run into an infinite loop but it will break with an `IllegalArgumentException` because we add the `Descriptor`'s name to the `msgDescriptorMapFull`, and if we encounter it again (via a recursive call), we throw an exception:


```java
private void addMessageType(Descriptor msgType, ...) {
        String msgTypeNameFull = msgType.getFullName();
        ...
        if (msgDescriptorMapFull.containsKey(msgTypeNameFull)) {
            throw new IllegalArgumentException(String.format(duplicateNameErrorMessage, msgTypeNameFull));
        }
        ...
        msgDescriptorMapFull.put(msgTypeNameFull, msgType);
        ...
        for (Descriptor nestedType : msgType.getNestedTypes()) {
            addMessageType(nestedType, msgTypeNameShort, msgDupes, enumDupes);
        ...
}
```

Not sure if this is an admissible behavior though.

Reply:
LGTM.